File: DesignPatterns/Structural/Bridge/Ruby/shape_client.rb
class RedColor
  def apply_color
    'Red.'
  end
end

class GreenColor
  def apply_color
    'Green.'
  end
end

class Shape
  attr_accessor :color

  def initialize(color)
    self.color = color
  end

  def create_shape
    raise NotImplementedError
  end
end

class Triangle < Shape
  def create_shape
    puts "Triangle filled with color #{@color.apply_color}"
  end
end

class Pentagon < Shape
  def create_shape
    puts "Pentagon filled with color #{@color.apply_color}"
  end
end

triangle = Triangle.new(RedColor.new)
triangle.create_shape
pentagon = Pentagon.new(GreenColor.new)
pentagon.create_shape

File: DesignPatterns/Structural/Bridge/Java/ShapeClient.java
interface Color {
  public void applyColor();
}

class RedColor implements Color {

  public void applyColor(){
    System.out.println(" Red.");
  }
}

class GreenColor implements Color {

  public void applyColor() {
    System.out.println(" Green.");
  }
}

abstract class Shape {
  protected Color color;

  public Shape(Color color){
    this.color = color;
  }

  abstract public void createShape();
}

class Triangle extends Shape {

  public Triangle(Color color){
    super(color);
  }

  public void createShape() {
    System.out.print("Triangle filled with color");
    color.applyColor();
  }
}

class Pentagon extends Shape {

  public Pentagon(Color color) {
    super(color);
  }

  public void createShape() {
    System.out.print("Pentagon filled with color");
    color.applyColor();
  }
}

public class ShapeClient {
  public static void main(String[] args) {
    Shape shape = new Triangle(new RedColor());
    shape.createShape();
    Shape shape1 = new Pentagon(new GreenColor());
    shape1.createShape();
  }
}